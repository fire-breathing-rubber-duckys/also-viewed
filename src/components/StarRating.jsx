/* eslint-disable react/no-array-index-key */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

const StarRating = ({ rating, ratingCount }) => {
  const [starsArr, setStarsArr] = useState([]);
  useEffect(() => {
    switch (rating) {
      case 0: setStarsArr([0, 0, 0, 0, 0]); break;
      case 0.5: setStarsArr([0.5, 0, 0, 0, 0]); break;
      case 1: setStarsArr([1, 0, 0, 0, 0]); break;
      case 1.5: setStarsArr([1, 0.5, 0, 0, 0]); break;
      case 2: setStarsArr([1, 1, 0, 0, 0]); break;
      case 2.5: setStarsArr([1, 1, 0.5, 0, 0]); break;
      case 3: setStarsArr([1, 1, 1, 0, 0]); break;
      case 3.5: setStarsArr([1, 1, 1, 0.5, 0]); break;
      case 4: setStarsArr([1, 1, 1, 1, 0]); break;
      case 4.5: setStarsArr([1, 1, 1, 1, 0.5]); break;
      case 5: setStarsArr([1, 1, 1, 1, 1]); break;
      default: break;
    }
  }, []);

  return (
    <div>
      {starsArr.map((star, i) => {
        if (star === 0) {
          return <i className="far fa-star" key={i} id="star" />;
        }
        if (star === 0.5) {
          return <i className="fas fa-star-half-alt" key={i} id="star" />;
        }
        if (star === 1) {
          return <i className="fas fa-star" key={i} id="star" />;
        }
        return undefined;
      })}
      <label className="item-rating-count">{ratingCount}</label>
    </div>
  );
};

StarRating.propTypes = {
  rating: PropTypes.number.isRequired,
  ratingCount: PropTypes.string.isRequired,
};

export default StarRating;
