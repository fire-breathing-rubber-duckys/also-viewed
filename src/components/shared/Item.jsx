import React from 'react';
import PropTypes from 'prop-types';

const Item = ({ item }) => (
  <div className="flex-item">
    <img src={item.photoURL} alt="" />
    <div>{item.brandName}</div>
    <div>{item.itemName}</div>
    <div>{item.itemRating}</div>
    <div>{item.ratingCount}</div>
    <div>{item.itemPrice}</div>
  </div>
);

Item.propTypes = {
  item: PropTypes.instanceOf(Object).isRequired,
};

export default Item;
