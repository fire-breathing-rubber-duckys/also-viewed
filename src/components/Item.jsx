/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import StarRating from './StarRating';

const ItemWrapper = styled.div`
  margin-left: 15px;
  margin-right: 15px;
  margin-bottom: 55px;
  position: relative;
  z-index: 1;
`;

const Item = ({ item }) => {
  const stringArr = item.itemName.split(' ');
  const ratingCount = ` (${item.ratingCount})`;
  const ip = item.itemPrice;
  const stringOne = [];
  const stringTwo = [];

  // Logic to filter item price by reg Vs sale item //
  let price;
  if (ip.length < 8 || ip[6] !== ' ') {
    price = <label className="item-price">{ip}</label>;
  } else {
    const priceArr = ip.split(' ');
    const priceOne = ` ${priceArr[0]}`;
    const priceTwo = ` ${priceArr[1]}`;
    price = (
      <div>
        <label className="item-sale-discount">{priceOne}</label>
        <label className="item-sale-price">{priceTwo}</label>
      </div>
    );
  }

  // Logic to filter the itemName to format in a nice fashion //
  if (stringArr[0].length > 8 && stringArr.length === 2) {
    stringOne.push(stringArr[0]);
    stringTwo.push(stringArr[1]);
  } else if (stringArr[0].length > 8 && stringArr.length === 3) {
    stringOne.push(stringArr[0]);
    stringTwo.push(stringArr[1], stringArr[2]);
  } else {
    for (let i = 0; i < stringArr.length; i += 1) {
      if (i < 2) {
        stringOne.push(stringArr[i]);
      } else {
        stringTwo.push(stringArr[i]);
      }
    }
  }

  return (
    <>
      <ItemWrapper>
        <img src={item.photoURL} alt="" className="also-viewed-img" />
        <div>
          <div className="item-name">
            <div className="item-brand">{item.brandName}</div>
            {stringOne.join(' ')}
            <br />
            {stringTwo.join(' ')}
          </div>
          <StarRating rating={item.itemRating} ratingCount={ratingCount} />
          {price}
        </div>
      </ItemWrapper>
    </>
  );
};

Item.propTypes = {
  item: PropTypes.instanceOf(Object).isRequired,
};

export default Item;
