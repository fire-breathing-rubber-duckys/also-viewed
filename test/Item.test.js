import React from 'react';
import { shallow } from 'enzyme';
import { expect as expectChai } from 'chai';
import renderer from 'react-test-renderer';
import Item from '../src/components/Item';
import StarRating from '../src/components/StarRating';

const item = {
  productId: 1,
  photoURL: 'https://www.rei.com/media/product/148949?size=280',
  brandName: 'Manduka',
  itemName: 'Lean Cork Block',
  itemRating: 5,
  itemPrice: '$16.00',
  ratingCount: 2,
};

it('Should render correctly', () => {
  const tree = renderer
    .create(<Item item={item} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('View Item Rendering', () => {
  const wrapper = shallow(<Item item={item} />);
  it('should render two (item-name, item-brand) divs', () => {
    expectChai(wrapper.find('.item-name')).to.be.lengthOf(1);
    expectChai(wrapper.find('.item-brand')).to.be.lengthOf(1);
  });
  it('should render one image', () => {
    expectChai(wrapper.find('img')).to.be.lengthOf(1);
  });
  it('should render StarRating once', ()=> {
    expectChai(wrapper.find(StarRating)).to.be.lengthOf(1);
  });
  it('should render one item-price dev', () => {
    expectChai(wrapper.find('.item-price')).to.be.lengthOf(1);
  });
});
