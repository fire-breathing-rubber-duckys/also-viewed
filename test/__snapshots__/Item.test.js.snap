// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
<div
  className="sc-bdfBwQ fECFQP"
>
  <img
    alt=""
    className="also-viewed-img"
    src="https://www.rei.com/media/product/148949?size=280"
  />
  <div>
    <div
      className="item-name"
    >
      <div
        className="item-brand"
      >
        Manduka
      </div>
      Lean Cork
      <br />
      Block
    </div>
    <div>
      <label
        className="item-rating-count"
      >
         (2)
      </label>
    </div>
    <label
      className="item-price"
    >
      $16.00
    </label>
  </div>
</div>
`;
